code/modules/mapping/reader.dm

```diff
@@ -367,33 +431,51 @@
 		// If our relative x is greater then X upper, well then we've gotta limit our expansion
 		var/delta = max(final_x - x_delta_with, 0)
 		final_x -= delta
-	if(final_x > world.maxx && !cropMap)
+	if(final_x > world.maxx && !crop_map)
 		if(new_z)
 			// Need to avoid improperly loaded area/turf_contents
-			world.increaseMaxX(final_x, max_zs_to_load = z_offset - 1)
+			world.increase_max_x(final_x, map_load_z_cutoff = z_offset - 1)
 		else
-			world.increaseMaxX(final_x)
+			world.increase_max_x(final_x)
 		expanded_x = TRUE
 
 	var/lowest_x = max(x_lower, 1 - x_relative_to_absolute)
 
+	// Amount we offset the grid zcrd to get the true zcrd
+	var/grid_z_offset = z_offset - 1
+	var/z_upper_set = z_upper < INFINITY
+	var/z_lower_set = z_lower > -INFINITY
+
 	// We make the assumption that the last block of turfs will have the highest embedded z in it
-	var/highest_z = last_column.zcrd + z_offset - 1 // Lets not just make a new z level each time we increment maxz
+	// true max zcrd
+	var/map_bounds_z_max = last_column.zcrd
+	var/z_upper_parsed = map_bounds_z_max + z_offset - 1
+	if(z_upper_set)
+		z_upper_parsed -= map_bounds_z_max - z_upper
+	if(z_lower_set)
+		var/offset_amount = z_lower - 1
+		z_upper_parsed -= offset_amount
+		grid_z_offset -= offset_amount
+
+	var/list/target_grid_sets = gridSets
+	if(z_lower_set || z_upper_set) // bounds are set, filter out gridsets for z levels we don't want
+		target_grid_sets = filter_grid_sets_based_on_z_bounds(z_lower, z_upper)
+
 	var/z_threshold = world.maxz
-	if(highest_z > z_threshold && cropMap)
-		for(var/i in z_threshold + 1 to highest_z) //create a new z_level if needed
+	if(z_upper_parsed > z_threshold && crop_map)
+		for(var/i in z_threshold + 1 to z_upper_parsed) //create a new z_level if needed
 			world.incrementMaxZ()
 		if(!no_changeturf)
 			WARNING("Z-level expansion occurred without no_changeturf set, this may cause problems when /turf/AfterChange is called")
 
-	for(var/datum/grid_set/gset as anything in gridSets)
+	for(var/datum/grid_set/gset as anything in target_grid_sets)
```

Comment:
this name makes me depressed

Reply:
target_grid_sets or gset;

If gset I blame you but can also make it better

Reply:
I can live with this I guess